File: packages/sdk-python/sentro/step.py
"""SentroStep -- tracks a single step within an agent run."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from .llm_call import SentroLlmCall
from .tool_call import SentroToolCall
from .transport import Transport


class SentroStep:
    """Represents a step within a run."""

    def __init__(
        self,
        transport: Transport,
        run_id: str,
        seq: int,
        content: str,
        *,
        capture_prompts: bool = False,
    ) -> None:
        self._transport = transport
        self._run_id = run_id
        self._seq = seq
        self._step_id = str(uuid.uuid4())
        self._capture_prompts = capture_prompts

        self._transport.send(
            {
                "type": "step.start",
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "runId": self._run_id,
                "stepId": self._step_id,
                "sequenceNumber": self._seq,
                "content": content,
            }
        )

    def tool_call(
        self,
        tool_name: str,
        *,
        input: dict[str, Any] | None = None,
    ) -> SentroToolCall:
        """Start a tool call within this step."""
        return SentroToolCall(
            self._transport,
            self._run_id,
            self._step_id,
            tool_name,
            input=input,
        )

    def trace_tool_call(
        self,
        tool_name: str,
        *,
        input: dict[str, Any] | None = None,
    ) -> SentroToolCall:
        """Start a tool call as a context manager."""
        return self.tool_call(tool_name, input=input)

    def llm_call(
        self,
        *,
        model: str,
        provider: str | None = None,
        messages: list[dict[str, str]] | None = None,
        temperature: float | None = None,
    ) -> SentroLlmCall:
        """Start an LLM call within this step."""
        return SentroLlmCall(
            self._transport,
            self._run_id,
            self._step_id,
            model=model,
            provider=provider,
            messages=messages,
            temperature=temperature,
            capture_prompts=self._capture_prompts,
        )

    def end(self) -> None:
        """End this step."""
        self._transport.send(
            {
                "type": "step.end",
                "timestamp": datetime.now(timezone.utc).isoformat(),
                "runId": self._run_id,
                "stepId": self._step_id,
                "sequenceNumber": self._seq,
            }
        )

    def __enter__(self) -> SentroStep:
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:  # type: ignore[no-untyped-def]
        self.end()

    async def __aenter__(self) -> SentroStep:
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:  # type: ignore[no-untyped-def]
        self.end()

File: packages/sdk/src/step.ts
import { Transport } from "./transport";
import { SentroToolCall } from "./tool-call";
import { SentroLlmCall } from "./llm-call";
import type { ToolCallOptions, LlmCallOptions } from "./types";

interface StepInternalOptions {
  capturePrompts: boolean;
}

export class SentroStep {
  private transport: Transport;
  private runId: string;
  private seq: number;
  private stepId: string;
  private capturePrompts: boolean;

  constructor(
    transport: Transport,
    runId: string,
    seq: number,
    content: string,
    internal: StepInternalOptions
  ) {
    this.transport = transport;
    this.runId = runId;
    this.seq = seq;
    this.stepId = crypto.randomUUID();
    this.capturePrompts = internal.capturePrompts;

    this.transport.send({
      type: "step.start",
      timestamp: new Date().toISOString(),
      runId: this.runId,
      stepId: this.stepId,
      seq: this.seq,
      content,
    });
  }

  toolCall(toolName: string, options?: ToolCallOptions): SentroToolCall {
    return new SentroToolCall(this.transport, this.runId, this.stepId, toolName, options);
  }

  async traceToolCall<T>(
    toolName: string,
    input: Record<string, unknown>,
    fn: (tc: SentroToolCall) => Promise<T>
  ): Promise<T> {
    const tc = this.toolCall(toolName, { input });
    try {
      const result = await fn(tc);
      await tc.end({ result });
      return result;
    } catch (err) {
      await tc.error(err instanceof Error ? err : new Error(String(err)));
      throw err;
    }
  }

  llmCall(options: LlmCallOptions): SentroLlmCall {
    return new SentroLlmCall(this.transport, this.runId, this.stepId, options, {
      capturePrompts: this.capturePrompts,
    });
  }

  async end(): Promise<void> {
    this.transport.send({
      type: "step.end",
      timestamp: new Date().toISOString(),
      runId: this.runId,
      stepId: this.stepId,
      seq: this.seq,
    });
  }
}
